File: Difficulty: Easy/Find length of Loop/find-length-of-loop.py
#User function Template for python3

'''
# Node Class
class Node:
    def __init__(self, data):   # data -> value stored in node
        self.data = data
        self.next = None
'''
class Solution:
    # Function to find the length of a loop in the linked list.
     def countNodesInLoop(self, head):
        #code here
        
        fast, slow = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                break
        else:
            return 0
            
        # now slow and fast must be inside the circle
        cnt = 1
        slow = slow.next
        while slow != fast:
            cnt += 1
            slow = slow.next
        return cnt
        #code here


#{ 
 # Driver Code Starts
class Node:

    def __init__(self, data):
        self.data = data
        self.next = None


def print_list(node):
    while node:
        print(node.data, end=' ')
        node = node.next
    print()


def loop_here(head, pos):
    if pos == 0:
        return

    walk = head
    for _ in range(1, pos):
        walk = walk.next

    tail = head
    while tail.next:
        tail = tail.next

    tail.next = walk


if __name__ == "__main__":
    import sys
    input = sys.stdin.read
    data = input().split('\n')
    t = int(data[0])
    index = 1
    for _ in range(t):
        arr = list(map(int, data[index].split()))
        k = int(data[index + 1])
        head = Node(arr[0])
        tail = head
        for value in arr[1:]:
            tail.next = Node(value)
            tail = tail.next
        loop_here(head, k)
        ob = Solution()
        res = ob.countNodesInLoop(head)
        print(res)
        print("~")
        index += 2

# } Driver Code Ends

File: Difficulty: Medium/Find length of Loop/find-length-of-loop.cpp
/*
class Node {
 public:
    int data;
    Node *next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};
*/

class Solution {
  public:
    int lengthOfLoop(Node *head) {
        // code here
        if(head==nullptr || head->next==nullptr)return 0;
        Node*slow=head; Node*fast=head; int len=0;
        while(fast and fast->next){
            slow=slow->next; fast=fast->next->next;
            if(slow==fast){
                len++;fast=fast->next;
                while(fast!=slow){fast=fast->next;len++;}
                return len;
            }
        }
        return len;
    }
};